python/pyspark/pandas/data_type_ops/base.py

```diff
@@ -0,0 +1,115 @@
+# limitations under the License.
+#
+
+from abc import ABCMeta, abstractmethod
+
+from pandas.api.types import CategoricalDtype
+
+from pyspark.sql.types import (
+    BooleanType,
+    DataType,
+    DateType,
+    FractionalType,
+    IntegralType,
+    StringType,
+    TimestampType,
+)
+
+from pyspark.pandas.typedef import Dtype
+
+
+class DataTypeOps(object, metaclass=ABCMeta):
+    """The base class for binary operations of pandas-on-Spark objects (of different data types)."""
+
+    def __new__(cls, dtype: Dtype, spark_type: DataType):
+        from pyspark.pandas.data_type_ops.boolean_ops import BooleanOps
+        from pyspark.pandas.data_type_ops.categorical_ops import CategoricalOps
+        from pyspark.pandas.data_type_ops.date_ops import DateOps
+        from pyspark.pandas.data_type_ops.datetime_ops import DatetimeOps
+        from pyspark.pandas.data_type_ops.num_ops import (
+            IntegralOps,
+            FractionalOps,
+        )
+        from pyspark.pandas.data_type_ops.string_ops import StringOps
+
+        if isinstance(dtype, CategoricalDtype):
+            return object.__new__(CategoricalOps)
+        elif isinstance(spark_type, FractionalType):
+            return object.__new__(FractionalOps)
+        elif isinstance(spark_type, IntegralType):
+            return object.__new__(IntegralOps)
+        elif isinstance(spark_type, StringType):
+            return object.__new__(StringOps)
+        elif isinstance(spark_type, BooleanType):
+            return object.__new__(BooleanOps)
+        elif isinstance(spark_type, TimestampType):
+            return object.__new__(DatetimeOps)
+        elif isinstance(spark_type, DateType):
+            return object.__new__(DateOps)
+        else:
+            raise TypeError("Type %s was not understood." % dtype)
+
+    def __init__(self, dtype: Dtype, spark_type: DataType):
+        self.dtype = dtype
+        self.spark_type = spark_type
+
+    @property
+    @abstractmethod
+    def pretty_name(self):
+        raise NotImplementedError()
+
+    def __add__(self, left, right):
+        raise TypeError("Addition can not be applied to %s." % self.pretty_name)
+
+    def __sub__(self, left, right):
+        raise TypeError("Subtraction can not be applied to %s." % self.pretty_name)
+
+    def __mul__(self, left, right):
+        raise TypeError("Multiplication can not be applied to %s." % self.pretty_name)
+
+    def __truediv__(self, left, right):
+        raise TypeError("True division can not be applied to %s." % self.pretty_name)
+
+    def __floordiv__(self, left, right):
+        raise TypeError("Floor division can not be applied to %s." % self.pretty_name)
+
+    def __mod__(self, left, right):
+        raise TypeError("Modulo can not be applied to %s." % self.pretty_name)
+
+    def __pow__(self, left, right):
+        raise TypeError("Exponentiation can not be applied to %s." % self.pretty_name)
+
+    def __radd__(self, left, right=None):
```

Comment:
Do we need `=None` here and in the following functions?

Reply:
`=None`s are removed. They are required when the Pull Request is against Koalas repo, though I am not clear about the reason now.